[Yeo Kai Jiun] iP

kaij77 opened the pull request:
# Duke

## Are you:
- [x] Unable to keep track of your **countless** tasks?
- [x] Familiar with **CLI**?
- [x] A **fast** typer?

If you ticked ~~all 3~~ at least 1 of the above, then Duke is for you!


## What does Duke do?
* Helps you keep track of your tasks!
* Categorize into different types of tasks.
   _1. ToDo
   2. Deadline
   3. Event_
* Mark tasks as done or unmark if not done.
* Search for a keyword to get all the matching tasks!


# Have fun!! 🥇 


This site was built using [GitHub Pages](https://pages.github.com/).

Commit:
Add Gradle support

kaij77 committed:
Implemented level 1

kaij77 committed:
Implemented Levels 2 and 3

kaij77 committed:
Implemented Level 4

kaij77 committed:
Implemented Level 5 with 3 different types of specific exceptions

kaij77 committed:
Implemented Level 5

kaij77 committed:
Implemented Level 6 and added a new Exception class to handle invalid deletions.

kaij77 committed:
Added javadocs to most public methods

kaij77 committed:
Changed DukeEmptyDescription to DukeInvalidDescriptionException in order to handle exceptions caused by missing deadlines or dates in Event or Deadline inputs.

kaij77 committed:
Update DukeInvalidDescriptionException.java

kaij77 committed:
Implemented Auto Text UI Testing

marcuspang commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+ */
+public class Deadline extends Task {
+
+    private String deadline;
```
Personal preference, but since `deadline` does not get updated, we can add the `final` keyword to it. This helps to signal to other developers that this field should be immutable (unless you are implementing a method to change this). 

Since this isn't in the coding standards, feel free to close this if you don't agree with me

marcuspang commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,56 @@
 public class Duke {
+
+    private UI ui;
+
+   public String logo = " ____        _        \n"
```
```suggestion
    private UI ui;

    public String logo = " ____        _        \n"
```

You have a missing space here

marcuspang commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,56 @@
+                }else {
+                    throw new DukeInvalidInputException();
+                }
+            } catch (DukeException e) {
```
I like how you abstract the different commands into different `ui` methods, makes thing very readable. I also like the fact that the exception is handled in `run` method so that the individual input methods can simply just throw `DukeException`s without worrying about printing it to the console

marcuspang commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,56 @@
+                    System.out.println(ui.list());
+                } else if (input.startsWith("mark") || input.startsWith("unmark")) {
+                    System.out.println(ui.editTask(input));
+                } else if (input.startsWith("todo ") || input.startsWith("event ") || input.startsWith("deadline ")){
+                    System.out.println(ui.addTask(input));
+                } else if (input.startsWith("delete")) {
+                    System.out.println(ui.deleteTask(input));
+                }else {
```
```suggestion
                } else if (input.startsWith("todo ") || input.startsWith("event ") || input.startsWith("deadline ")) {
                    System.out.println(ui.addTask(input));
                } else if (input.startsWith("delete")) {
                    System.out.println(ui.deleteTask(input));
                } else {
```
>  White space within a statement

You are missing some whitespaces here

marcuspang commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,29 @@
+ */
+public class Event extends Task {
+
+    private String dateTime;
```
Ditto for `final`, feel free to close

marcuspang commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Task {
```
>  Write descriptive header comments for all public classes/methods.

Missing javadoc for `Task` and methods, but I believe you will address this while doing the `A-JavaDoc` level

ShummyOwnzYou replied:
Perhaps you could put all your Task classes in a different directory as this directory is getting a bit too congested

marcuspang commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public boolean getIsDone() {
+        return isDone;
+    }
+
+    public void mark() {
+        this.isDone = true;
+    }
+
+    public void unmark() {
+        this.isDone = false;
```
I think it'll be good to standardise where you want to use `this.isDone` or `isDone` in these methods. But this isn't really a violation since the use of `this` doesn't affect the functionality in these methods. 

Feel free to close this if you disagree

marcuspang commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+     * @return String which content depends on whether the current task list is empty
+     */
+    public String list() {
+
```
```suggestion
```
Unnecessary empty line

marcuspang commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+     * @return String telling the user what task was added and the number of current tasks.
+     * @throws DukeInvalidDescriptionException if the description of the given is empty.
+     */
+    public String addTask(String str) throws DukeInvalidDescriptionException {
```
I like how you abstracted the commands into different methods, which will definitely help you in the subsequent methods. 👍

marcuspang commented on text-ui-test/runtest.txt:
```diff
@@ -0,0 +1,21 @@
+@ECHO OFF
```
This is separate, but I think you accidentally committed this file

marcuspang reviewed (commented):
There were some minor coding violations, but they should be addressed easily with your IDE (just be more careful!) and in future levels (especially regarding the lack of javadoc for some classes/methods).

Overall, I think you did a good job so keep it up! 👍

ShummyOwnzYou commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * A DukeException is thrown if there is any error related to Duke the chatbot.
+ */
+public class DukeException extends Exception {
```
Perhaps you could put all your DukeException classes in a different directory as this directory is getting a bit too congested

ShummyOwnzYou commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class UI {
```
Perhaps `UI.java` should be renamed to `Ui.java`, as stated in A-MoreOOP level, to comply with coding standards as well.

ShummyOwnzYou commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+    private Scanner sc;
+    private List<Task> taskList;
+
+    private static String initGreeting = "Hello! I'm Duke \nWhat can I do for you?";
```
Since this `String` does not change throughout the program, perhaps consider adding the `final` keyword.

ShummyOwnzYou commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+        } else {
+            String temp = "";
+            for (int i = 0; i < taskList.size(); i++) {
+                int curr = i + 1;
```
Perhaps rename `curr` to `index` or `currIndex` to make the variable more descriptive to the reader, without needing to read more code to figure our what `curr` indicates.

ShummyOwnzYou commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+     * @return String containing the farewell message
+     */
+    public String bye() {
+        return "Bye. Hope to see you again soon!";
```
Is there any reason why the `greet()` method returns the `initGreeting` attribute while this `bye()` method just returns the message as a String? Perhaps consider using the same logic for both methods to be consistent and not confuse readers.

ShummyOwnzYou commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+     * @throws DukeMarkException if the task has already been marked as done.
+     */
+    public String markTask(int id) throws DukeMarkException {
+        Task curr = taskList.get(id - 1);
```
Same naming issue as above.

ShummyOwnzYou commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+     * @throws DukeMarkException if the task has not been marked as done (already un-marked).
+     */
+    public String unmarkTask(int id) throws DukeMarkException {
+        Task curr = taskList.get(id - 1);
```
Same naming issue as above.

ShummyOwnzYou commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,56 @@
+                    System.out.println(ui.list());
+                } else if (input.startsWith("mark") || input.startsWith("unmark")) {
+                    System.out.println(ui.editTask(input));
+                } else if (input.startsWith("todo ") || input.startsWith("event ") || input.startsWith("deadline ")){
```
`){` should be `) {` at the end of the line, coding standard violation.

ShummyOwnzYou reviewed (commented):
Overall, some slight issues and some overdue tasks to be done, keep up the effort.

sh4nH commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+     *
+     * @return String containing the farewell message
+     */
+    public String bye() {
```
Why does the bye() method return a String? Maybe you could create a static final initBye variable to store the message.

sh4nH commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,174 @@
+        this.taskList = new ArrayList<Task>();
+    }
+
+    public String greet() {
```
Maybe you could use a more descriptive method name such as 'giveGreeting()'?

sh4nH commented on src/main/java/DukeInvalidInputException.java:
```diff
@@ -0,0 +1,13 @@
+     * caught.
+     */
+    public DukeInvalidInputException() {
+        super("☹ OOPS!!! I'm sorry, but I don't know what that means D:");
```
I like how you made very nice error messages. Improves the user experience!

sh4nH reviewed (commented):
Overall your code is very well done. There are a few coding standard violations, which have been pointed out by other reviewers as well. Consider organising your classes into packages.

kaij77 committed:
level 8

kaij77 committed:
Update Duke.java

kaij77 committed:
Implemented level 8

kaij77 committed:
Implemented A-MoreOOP

kaij77 committed:
test

kaij77 committed:
Organized files into
1. duke
2. duke.command
3.duke.exception
4. duke.task
packages

kaij77 committed:
added duke package

kaij77 committed:
Merge branch 'master' of https://github.com/kaij77/ip

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Ui.java

kaij77 committed:
added duke.command package

kaij77 committed:
added duke.exception and duke.task packages

kaij77 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

kaij77 committed:
Added JUnit tests for ListCommand and ToDo classes

kaij77 committed:
added JUnit tests

kaij77 committed:
Added the neccessary dependencies to build.gradle and built JAR file

kaij77 committed:
Added JavaDoc

kaij77 committed:
Implemented Level 9, Find.

kaij77 committed:
Implemented GUI for Duke

kaij77 committed:
No assertions are used in the code.

Assertions should be added to help identify any possible bugs
and verify them during runtime.

Added assertions in Parser.java and Ui.java

This is to check for any runtiem errors in certain methods belonging
to these classes.

kaij77 committed:
Not all parts of the code follow code quality standards.

ALl or most parts should follow these standards so that the code is
more readable and minimizes unsafe practices.

Some parts of the code in Duke.java is edited to make it more
readable. Unused exception classes were deleted to minimize unsafe
practices.

These were the parts of code currently identified to not be up to
code quality standards, and were hence edited accordingly.

kaij77 committed:
Merge pull request #3 from kaij77/branch-A-Assertions

No assertions are used in the code.

kaij77 committed:
Merge pull request #2 from kaij77/branch-A-CodeQuality

Branch a code quality

kaij77 committed:
Having notes about each task is not supported yet.

The user may want to add more notes on top of the task description.
Hence there is a need to support this system.

Implemented adding, editing and deleting of notes for each task.

This is to allow the user to add notes to each task and edit or
delete them as they wish.

User command syntax is as follows
1. Add or edit note
note /task (task number) /edit (new note)

2. Delete task
note /task (task number) /delete

kaij77 committed:
Added credit

kaij77 committed:
Changed Duke and User icons

kaij77 committed:
Fixed a small bug in Event.toString() and added Ui.png to the
docs folder.

kaij77 committed:
Wrote part of the UG in README.md

kaij77 committed:
Fixed bug where "bye" does not cause Duke to close, and edited README.md

kaij77 committed:
Added images to README.md

kaij77 committed:
Update README.md

kaij77 committed:
Finalized User Guide in README.md

kaij77 committed:
Update README.md

kaij77 committed:
Fix merge conflicts

kaij77 committed:
Edit README

kaij77 committed:
Fix Duke for re-submission

kaij77 committed:
Edit README.md

kaij77 committed:
Add images to correct folder

kaij77 committed:
Fix bug where JAR file cannot be opened
